src/mip/cg_strengthening.jl

```diff
@@ -0,0 +1,142 @@
+"""
+Perform Chavatal-Gomory strengthening of inequalities
+
+We apply the Chvatal-Goromory procedure to constraints that
+satisfy the following criterions:
+1. All variables with nonzero coefficient must be integral
+2. The constraints are not ranged
+3. If the constraints is of the form ax ⩾ b, then every
+   positive coefficient variables must have finite lowerbound
+   negative coefficient variables must have finite upperbound
+
+   If the constraints is of the from ax ⩽ b, then every
+   positive coefficient variables must have finite upperbound
+   negative coefficient variables must have finite lowerbound
+
+For example, let aᵢx ⩾ bᵢ is a constraint where aᵢ ⩾ 0, and x
+is integer variables. If there exist s > 0 such that
+⌈aᵢⱼ s⌉ bᵢ / ⌈bᵢ s⌉ ⩽ aᵢⱼ
+and there is k such that
+⌈aᵢₖ s⌉ bᵢ / ⌈bᵢ s⌉ < aᵢₖ (*)
+then we will update aᵢx ⩾ bᵢ into ⌈aᵢ s⌉ x ⩾ ⌈bᵢ s⌉
+
+We find such a number s heuristicaly i.e, we try different
+values from the following set
+{1, t/a_max, t/a_min, (2t-1)/(2a_min)| t = 1,...,search_range}
+where a_max = sup(|aᵢ|), a_min = inf(|aᵢ|) and search_range
+is set to be 5 by default
+
+"""
+function cg_strengthening!(ps::PresolveData{T}, search_range::Int = 5) where {T}
+    ps.pb0.is_continuous && return nothing
+
+    i_index = zeros(Int, ps.pb0.nvar)
```

Comment:
What is this used for? It seems to be counting the sparsity of each column. But, when you pass it to `cg_inequality` below it will be only partially populated.

Reply:
It is used to update ps.pb0.acols. Because we store as a list of nzind and nzval, I need a way  to keep track of the index.

Reply:
This vector is only partially populated when you call cg_inequality!, though. Imagine that i=1. In this case, i_index only counts the sparsity from the first row and nothing else. Is this sufficient to update the columns?

Per my other comment, I think you instead want to defer updating the columns until after you apply the strengthenings to all rows (and this is really a generic operation that you can abstract out and use across multiple reductions)